project-code/core/common/src/main/scala/play/core/server/servlet/RequestHandler.scala

```diff
@@ -71,26 +75,33 @@ trait HttpServletRequestHandler extends RequestHandler {
    */
   protected def onHttpResponseComplete(): Unit
 
-  protected def feedBodyParser(bodyParser: Iteratee[Array[Byte], Result]): Future[Result] = {
-    // default synchronous blocking body enumerator
-
-    // FIXME this default body enumerator reads the entire stream in memory
-    // uploading a lot of data can lead to OutOfMemoryException
-    // For more details: https://github.com/dlecan/play2-war-plugin/issues/223
-    val bodyEnumerator = getHttpRequest().getRichInputStream.fold(Enumerator.eof[Array[Byte]]) { is =>
-      val output = new java.io.ByteArrayOutputStream()
-      val buffer = new Array[Byte](1024 * 8)
-      var length = is.read(buffer)
-      while(length != -1){
-        output.write(buffer, 0, length)
-        length = is.read(buffer)
-      }
-      Enumerator(output.toByteArray) andThen Enumerator.eof
+  /** Create the source for the request body */
+  protected def convertRequestBody()(implicit mat: Materializer): Option[Source[ByteString, Any]] = {
+    getHttpRequest().getRichInputStream.map { is ⇒
+      StreamConverters.fromInputStream(() ⇒ is)
```

Comment:
This is uses blocking IO, it's not the right approach, and it's a regression of the current behaviour, which uses the Servlet 3.1 non blocking IO APIs.

The hard work of converting Servlet 3.1 non blocking IO to reactive streams has already been done here:

https://github.com/jroper/reactive-streams-servlet

You just need to copy that code into this project (specifically these three classes: https://github.com/jroper/reactive-streams-servlet/tree/master/reactive-streams-servlet/src/main/java/org/reactivestreams/servlet), and then you can plumb the servlet input stream to Akka streams with non blocking IO.

Reply:
How about including this as a dependency to the project? I would prefer not to copy, but reuse your awesome work. :)

Reply:
I was able to use reactive-streams-servlet. Can you please verify if this is correct?